core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java

```diff
@@ -0,0 +1,153 @@
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class GracefulShutdownCoordinatorHandler
+        implements GracefulShutdownHandler
+{
+    private static final Logger log = Logger.get(GracefulShutdownCoordinatorHandler.class);
+    private static final Duration LIFECYCLE_STOP_TIMEOUT = new Duration(30, SECONDS);
+
+    private final ScheduledExecutorService shutdownHandler = newSingleThreadScheduledExecutor(threadsNamed("shutdown-handler-%s"));
+    private final ExecutorService lifeCycleStopper = newSingleThreadExecutor(threadsNamed("lifecycle-stopper-%s"));
+    private final LifeCycleManager lifeCycleManager;
+    private final StartupStatus status;
+    private final QueryManager queryManager;
+    private final ShutdownAction shutdownAction;
+    private final Duration gracePeriod;
+
+    @GuardedBy("this")
+    private boolean shutdownRequested;
+
+    @Inject
+    public GracefulShutdownCoordinatorHandler(
+            StartupStatus status,
+            QueryManager queryManager,
+            ServerConfig serverConfig,
+            ShutdownAction shutdownAction,
+            LifeCycleManager lifeCycleManager)
+    {
+        this.status = requireNonNull(status, "status is null");
+        this.queryManager = requireNonNull(queryManager, "queryManager is null");
+        this.shutdownAction = requireNonNull(shutdownAction, "shutdownAction is null");
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+        this.gracePeriod = requireNonNull(serverConfig, "serverConfig is null").getGracePeriod();
+    }
+
+    @Override
+    public synchronized void requestShutdown()
+    {
+        log.info("Shutdown requested");
+        if (shutdownRequested) {
+            return;
+        }
+        shutdownRequested = true;
+
+        status.shutdownStarted();
+        // wait for a grace period to start the shutdown sequence
+        shutdownHandler.schedule(this::shutdown, gracePeriod.toMillis(), MILLISECONDS);
+    }
+
+    private void shutdown()
+    {
+        log.info("Shutting down");
+        List<BasicQueryInfo> activeQueries = getActiveQueries();
+
+        // At this point no new queries should be queued by the coordinator.
+        // Wait for all remaining queries to finish.
+        while (!activeQueries.isEmpty()) {
+            CountDownLatch countDownLatch = new CountDownLatch(activeQueries.size());
+
+            for (BasicQueryInfo queryInfo : activeQueries) {
+                queryManager.addStateChangeListener(queryInfo.getQueryId(), newState -> {
```

Comment:
Is a race condition possible here, when an active query will be finished before attaching this change listener?

Reply:
no, addStateChangeListener will always fire at least once with the last observed state afair

Reply:
```
// fire state change listener with the current state
        // always fire listener callbacks from a different thread
        safeExecute(() -> fireStateChangedListener(currentState, stateChangeListener));
```